catalog/dags/providers/provider_api_scripts/cc_mixter.py

```diff
@@ -0,0 +1,263 @@
+"""
+Content Provider:       ccMixter
+
+ETL Process:            Use the API to identify all CC licensed media.
+
+Output:                 TSV file containing the media and the
+                        respective meta-data.
+
+Notes:                  Documentation: https://ccmixter.org/query-api
+                        ccMixter sends bad JSON and extremely huge headers, both
+                        of which need workarounds that are handled by this DAG.
+"""
+
+import json
+import logging
+import re
+from typing import Literal
+
+from common import constants
+from common.licenses import get_license_info
+from common.loader import provider_details as prov
+from common.requester import DelayedRequester
+from providers.provider_api_scripts.provider_data_ingester import ProviderDataIngester
+
+
+logger = logging.getLogger(__name__)
+
+JSON_OCTALS = re.compile(r":\s*0(?P<num>\d+)\s*(?P<sep>[,}])")
+
+
+class CcMixterDelayedRequester(DelayedRequester):
+    """
+    WORKAROUND!
+
+    ccMixter sends bad JSON, including numbers with a leading 0 (observed some
+    such cases with "bpm" field). This makes the JSON invalid and raises decode
+    errors.
+
+    This class extends ``DelayedRequester`` to supply a custom JSON decoding
+    step where we perform a text substitution first and then parse the JSON.
+    """
+
+    def _get_json(self, response):
+        raw_json = response.text
+        cleaned_json = JSON_OCTALS.sub(r":\g<num>\g<sep>", raw_json)
+        if cleaned_json == raw_json:
+            logger.info("JSON was clean, no substitutions were made.")
+        else:
+            logger.warning("JSON had bad octals, substitutions were made.")
+        try:
+            response_json = json.loads(cleaned_json)
+        except json.JSONDecodeError as e:
+            logger.warning(f"Could not get response_json.\n{e}")
+            response_json = None
```

Comment:
Ideally there is no point in retrying the request because it will be the same response and raise the same error till it exhausts all tries. I just used this as the minimum change to ``DelayedRequested`` so that it can workaround ccMixter's issues.

Reply:
I ran into this case in my tests. For confirmation, do we want it to fail if the json can't be decoded or we want it to continue? In the first case, would that mean the DAG requires modifications?

Reply:
I would think that it makes more sense for the DAG to fail when the JSON is corrupt because then it's not a problem with a specific record (that we can ignore and move on) but rather with hundreds (an entire page's worth) of records.

Reply:
It ended up with `"offset": 26800`. There is no much more info in the stack trace.

```python
[2023-12-15, 02:06:51 UTC] {cc_mixter.py:49} WARNING - JSON had bad octals, substitutions were made.
[2023-12-15, 02:06:51 UTC] {cc_mixter.py:58} WARNING - Could not get response_json.
Expecting ',' delimiter: line 1 column 114472 (char 114471)
[2023-12-15, 02:06:51 UTC] {requester.py:151} WARNING - Bad response_json:  None
[2023-12-15, 02:06:51 UTC] {requester.py:152} WARNING - Retrying:
_get_response_json(
    https://ccmixter.org/api/query/,
    {'format': 'json', 'limit': 100, 'offset': 26800},
    retries=0)
[2023-12-15, 02:06:52 UTC] {cc_mixter.py:49} WARNING - JSON had bad octals, substitutions were made.
[2023-12-15, 02:06:52 UTC] {cc_mixter.py:58} WARNING - Could not get response_json.
Expecting ',' delimiter: line 1 column 114472 (char 114471)
[2023-12-15, 02:06:52 UTC] {requester.py:151} WARNING - Bad response_json:  None
[2023-12-15, 02:06:52 UTC] {requester.py:152} WARNING - Retrying:
_get_response_json(
    https://ccmixter.org/api/query/,
    {'format': 'json', 'limit': 100, 'offset': 26800},
    retries=-1)
[2023-12-15, 02:06:52 UTC] {requester.py:141} ERROR - No retries remaining.  Failure.
[2023-12-15, 02:06:52 UTC] {media.py:233} INFO - Writing 31 lines from buffer to disk.
[2023-12-15, 02:06:52 UTC] {provider_data_ingester.py:505} INFO - Committed 26631 records
[2023-12-15, 02:06:52 UTC] {taskinstance.py:1937} ERROR - Task failed with exception
providers.provider_api_scripts.provider_data_ingester.IngestionError: Retries exceeded
query_params: {"format": "json", "limit": 100, "offset": 26800}
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 192, in execute
    return_value = self.execute_callable()
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 209, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/opt/airflow/catalog/dags/providers/factory_utils.py", line 55, in pull_media_wrapper
    data = ingester.ingest_records()
  File "/opt/airflow/catalog/dags/providers/provider_api_scripts/provider_data_ingester.py", line 269, in ingest_records
    raise error from ingestion_error
  File "/opt/airflow/catalog/dags/providers/provider_api_scripts/provider_data_ingester.py", line 231, in ingest_records
    batch, should_continue = self.get_batch(query_params)
  File "/opt/airflow/catalog/dags/providers/provider_api_scripts/provider_data_ingester.py", line 392, in get_batch
    response_json = self.get_response_json(query_params)
  File "/opt/airflow/catalog/dags/providers/provider_api_scripts/provider_data_ingester.py", line 413, in get_response_json
    return self.delayed_requester.get_response_json(
  File "/opt/airflow/catalog/dags/common/requester.py", line 159, in get_response_json
    response_json = self.get_response_json(
  File "/opt/airflow/catalog/dags/common/requester.py", line 159, in get_response_json
    response_json = self.get_response_json(
  File "/opt/airflow/catalog/dags/common/requester.py", line 159, in get_response_json
    response_json = self.get_response_json(
  [Previous line repeated 1 more time]
  File "/opt/airflow/catalog/dags/common/requester.py", line 142, in get_response_json
    raise RetriesExceeded("Retries exceeded")
common.requester.RetriesExceeded: Retries exceeded
[2023-12-15, 02:06:52 UTC] {taskinstance.py:1400} INFO - Marking task as FAILED. dag_id=cc_mixter_workflow, task_id=ingest_data.pull_audio_data, execution_date=20231215T020212, start_date=20231215T020218, end_date=20231215T020652
[2023-12-15, 02:06:52 UTC] {slack.py:320} INFO - 
*DAG*: `cc_mixter_workflow`
*Task*: `ingest_data.pull_audio_data`
*Logical Date*: 2023-12-15T02:02:12Z
*Log*: http://localhost:8080/log?execution_date=2023-12-15T02%3A02%3A12%2B00%3A00&task_id=ingest_data.pull_audio_data&dag_id=cc_mixter_workflow&map_index=-1
*Exception*: Retries exceeded
*Exception Type*: `common.requester.RetriesExceeded`
[2023-12-15, 02:06:52 UTC] {base.py:73} INFO - Using connection ID 'slack_alerts' for task execution.
[2023-12-15, 02:06:52 UTC] {standard_task_runner.py:104} ERROR - Failed to execute job 97 for task ingest_data.pull_audio_data (Retries exceeded; 335)
[2023-12-15, 02:06:52 UTC] {local_task_job_runner.py:228} INFO - Task exited with return code 1
[2023-12-15, 02:06:53 UTC] {taskinstance.py:2778} INFO - 2 downstream tasks scheduled from follow-on schedule check
```